File: journal_django/frontend/admin-src/src/lib/labels.ts
import type { EnrollmentStatus, LessonType, RegistryStatus } from './types';

// ===== Enrollment status =====
// Подписи в Title Case — используются и в badge, и в формах, и в фильтрах.
// Бэк хранит коды (enrolled/frozen/not_enrolled/declined), UI показывает подписи.

export const ENROLLMENT_STATUS_LABELS: Record<EnrollmentStatus, string> = {
  enrolled:     'Учится',
  not_enrolled: 'Не учится',
  frozen:       'Заморожен',
  declined:     'Отказался',
};

export const ENROLLMENT_STATUS_OPTIONS: { value: EnrollmentStatus; label: string }[] =
  (Object.entries(ENROLLMENT_STATUS_LABELS) as [EnrollmentStatus, string][])
    .map(([value, label]) => ({ value, label }));

// ===== Реестр куратора — статус ученика =====
// Подписи бейджа и строк сигналов. Коды приходят с бэка (registry_service.classify).

export const REGISTRY_STATUS_LABELS: Record<RegistryStatus, string> = {
  closed:  'Пакет закрыт',
  ending:  'Заканчивается',
  idle:    'Простой',
  no_plan: 'Нет плана',
  ok:      'Активен',
};

// ===== Lesson type =====

export const LESSON_TYPE_LABELS: Record<LessonType, string> = {
  regular:      'обычный',
  substitution: 'замена',
  reschedule:   'перенос',
};

export const LESSON_TYPE_OPTIONS: { value: LessonType; label: string }[] =
  (Object.entries(LESSON_TYPE_LABELS) as [LessonType, string][])
    .map(([value, label]) => ({ value, label }));

// ===== Changelog: операции журнала изменений =====
// Ключи — из apps/changelog/labels.py (бэкенд выводит их из method+url).

// Короткие названия действий — колонка «Действие» и dropdown «Все действия».
export const CHANGELOG_OPERATION_LABELS: Record<string, string> = {
  'group.create':                  'Новая группа',
  'group.update':                  'Правка группы',
  'group.delete':                  'Группа в архив',
  'group.schedule_change':         'Смена расписания',
  'plan.generate':                 'Генерация плана',
  'plan.permanent_change':         'Смена расписания (план)',
  'plan.change_teacher_permanent': 'Смена преподавателя',
  'plan.change_teacher':           'Замена преподавателя',
  'plan.extra':                    'Доп. занятие',
  'plan.reschedule':               'Перенос урока',
  'plan.cancel':                   'Отмена урока',
  'direction.create':              'Новое направление',
  'direction.update':              'Правка направления',
  'direction.delete':              'Направление в архив',
  'teacher.create':                'Новый преподаватель',
  'teacher.update':                'Правка преподавателя',
  'teacher.delete':                'Преподаватель в архив',
  'student.create':                'Новый ученик',
  'student.update':                'Правка ученика',
  'student.delete':                'Ученик в архив',
  'student.status':                'Смена статуса ученика',
  'student.resume':                'Выход из заморозки',
  'discount.create':               'Новая скидка',
  'discount.update':               'Правка скидки',
  'discount.delete':               'Скидка в архив',
  'membership.create':             'Зачисление',
  'membership.update':             'Правка членства',
  'membership.delete':             'Отчисление',
  'payment.create':                'Оплата',
  'payment.delete':                'Удаление оплаты',
  'lesson.submit':                 'Проведение урока',
  'lesson.create':                 'Создание урока',
  'lesson.update':                 'Правка урока',
  'lesson.delete':                 'Удаление урока',
  'lesson.attendance_update':      'Правка посещаемости',
  'extra_lesson.create':           'Назначение доп.урока',
  'extra_lesson.cancel':           'Отмена доп.урока',
  'extra_lesson.delete':           'Удаление доп.урока',
  'extra_lesson.record':           'Проведение доп.урока',
  'payroll.update':                'Правка начисления',
  'settings.update':               'Настройки',
  'account.create':                'Новая учётка',
  'account.update':                'Правка учётки',
  'account.delete':                'Учётка выключена',
  'account.reset_password':        'Сброс пароля',
  'account.reset_2fa':             'Сброс 2FA',
  'account.invite_create':         'Invite-ссылка',
  'account.invite_revoke':         'Отзыв invite',
  'account.invite_accept':         'Активация учётки',
  'account.twofa_enable':          'Включение 2FA',
  'account.twofa_disable':         'Выключение 2FA',
  'changelog.revert':              'Откат',
  other:                           'Другое действие',
};

export const CHANGELOG_OPERATION_OPTIONS: { value: string; label: string }[] =
  Object.entries(CHANGELOG_OPERATION_LABELS).map(([value, label]) => ({ value, label }));

// ===== Changelog: сущности (registry.py → entity) =====

export const CHANGELOG_ENTITY_LABELS: Record<string, string> = {
  direction:      'Направление',
  teacher:        'Преподаватель',
  student:        'Ученик',
  discount:       'Скидка',
  settings:       'Настройки',
  account:        'Учётка',
  group:          'Группа',
  schedule_slot:  'Слот расписания',
  membership:     'Членство',
  planned_lesson: 'Плановое занятие',
  lesson:         'Урок',
  attendance:     'Посещаемость',
  payment:        'Оплата',
  payroll:        'Начисление',
  extra_lesson_assignment:  'Доп.урок (назначение)',
  extra_lesson_participant: 'Доп.урок (участник)',
};

export const CHANGELOG_ENTITY_OPTIONS: { value: string; label: string }[] =
  Object.entries(CHANGELOG_ENTITY_LABELS).map(([value, label]) => ({ value, label }));

// ===== Продления: стадии воронки (fallback, если под рукой только key —
// основной источник истины — stage_label/label с бэка) =====

export const RENEWAL_STAGE_LABELS: Record<string, string> = {
  no_lesson_yet: 'Не было урока', lesson_1: 'Урок 1', lesson_2: 'Урок 2', lesson_3: 'Урок 3',
  awaiting_payment: 'Ждём оплату', awaiting_renewal: 'Ждём продление', thinking: 'Думает',
  frozen: 'Заморожен', ignoring: 'Игнорит', renewed: 'Продлён', churned: 'Ушёл',
};

// Причины ухода ученика (диалог закрытия сделки «Ушёл», reason_code сделки).
export const RENEWAL_LOST_REASON_LABELS: Record<string, string> = {
  price: 'Не устроила цена',
  schedule: 'Не подошло расписание',
  lost_interest: 'Потерял интерес',
  relocation: 'Переезд',
  other: 'Другое',
};

File: journal_django/apps/changelog/labels.py
"""
Метки операций журнала: (HTTP-метод, url) → машинный ключ операции.

Русские названия — на фронте (lib/labels.ts). Ключ 'other' — fallback
для незамапленных мутаций: журнал остаётся читаемым, событие не теряется.
Порядок правил важен: более специфичные пути выше.
"""
from __future__ import annotations

import re

# (method, compiled regex, operation)
RULES: list[tuple[str, re.Pattern, str]] = [
    # scheduling (план занятий) — до generic groups-правил
    ('POST', re.compile(r'^/api/admin/groups/\d+/plan/generate$'), 'plan.generate'),
    ('POST', re.compile(r'^/api/admin/groups/\d+/plan/permanent-change$'), 'plan.permanent_change'),
    ('POST', re.compile(r'^/api/admin/groups/\d+/plan/change-teacher-permanent$'), 'plan.change_teacher_permanent'),
    ('POST', re.compile(r'^/api/admin/groups/\d+/plan/extra$'), 'plan.extra'),
    ('POST', re.compile(r'^/api/admin/groups/\d+/plan/\d+/reschedule$'), 'plan.reschedule'),
    ('POST', re.compile(r'^/api/admin/groups/\d+/plan/\d+/change-teacher$'), 'plan.change_teacher'),
    ('POST', re.compile(r'^/api/admin/groups/\d+/plan/\d+/cancel$'), 'plan.cancel'),
    # groups
    ('POST', re.compile(r'^/api/admin/groups/\d+/schedule-change$'), 'group.schedule_change'),
    ('POST', re.compile(r'^/api/admin/groups$'), 'group.create'),
    ('PATCH', re.compile(r'^/api/admin/groups/\d+$'), 'group.update'),
    ('DELETE', re.compile(r'^/api/admin/groups/\d+$'), 'group.delete'),
    # справочники
    ('POST', re.compile(r'^/api/admin/directions$'), 'direction.create'),
    ('PATCH', re.compile(r'^/api/admin/directions/\d+$'), 'direction.update'),
    ('DELETE', re.compile(r'^/api/admin/directions/\d+$'), 'direction.delete'),
    ('POST', re.compile(r'^/api/admin/teachers$'), 'teacher.create'),
    ('PATCH', re.compile(r'^/api/admin/teachers/\d+$'), 'teacher.update'),
    ('DELETE', re.compile(r'^/api/admin/teachers/\d+$'), 'teacher.delete'),
    ('POST', re.compile(r'^/api/admin/students$'), 'student.create'),
    ('PATCH', re.compile(r'^/api/admin/students/\d+$'), 'student.update'),
    ('DELETE', re.compile(r'^/api/admin/students/\d+$'), 'student.delete'),
    ('POST', re.compile(r'^/api/admin/discounts$'), 'discount.create'),
    ('PATCH', re.compile(r'^/api/admin/discounts/\d+$'), 'discount.update'),
    ('DELETE', re.compile(r'^/api/admin/discounts/\d+$'), 'discount.delete'),
    # memberships
    ('POST', re.compile(r'^/api/admin/memberships$'), 'membership.create'),
    ('PATCH', re.compile(r'^/api/admin/memberships/\d+$'), 'membership.update'),
    ('DELETE', re.compile(r'^/api/admin/memberships/\d+$'), 'membership.delete'),
    # payments (immutable: только create/delete)
    ('POST', re.compile(r'^/api/admin/payments$'), 'payment.create'),
    ('DELETE', re.compile(r'^/api/admin/payments/\d+$'), 'payment.delete'),
    # lessons
    ('PATCH', re.compile(r'^/api/admin/lessons/\d+/attendance/\d+$'), 'lesson.attendance_update'),
    ('POST', re.compile(r'^/api/admin/lessons$'), 'lesson.create'),
    ('PATCH', re.compile(r'^/api/admin/lessons/\d+$'), 'lesson.update'),
    ('DELETE', re.compile(r'^/api/admin/lessons/\d+$'), 'lesson.delete'),
    # payroll / settings
    ('PATCH', re.compile(r'^/api/admin/payroll/\d+$'), 'payroll.update'),
    ('PUT', re.compile(r'^/api/admin/settings$'), 'settings.update'),
    # accounts
    ('POST', re.compile(r'^/api/admin/accounts/\d+/reset-password$'), 'account.reset_password'),
    ('POST', re.compile(r'^/api/admin/accounts/\d+/reset-2fa$'), 'account.reset_2fa'),
    ('POST', re.compile(r'^/api/admin/accounts/\d+/invite/revoke$'), 'account.invite_revoke'),
    ('POST', re.compile(r'^/api/admin/accounts/\d+/invite$'), 'account.invite_create'),
    ('POST', re.compile(r'^/api/admin/accounts$'), 'account.create'),
    ('PATCH', re.compile(r'^/api/admin/accounts/\d+$'), 'account.update'),
    ('DELETE', re.compile(r'^/api/admin/accounts/\d+$'), 'account.delete'),
    # renewals (стадии-справочник — до generic deal-правил)
    ('POST', re.compile(r'^/api/admin/renewals/stages/reorder$'), 'renewal.stage_reorder'),
    ('POST', re.compile(r'^/api/admin/renewals/stages$'), 'renewal.stage_create'),
    ('PATCH', re.compile(r'^/api/admin/renewals/stages/\d+$'), 'renewal.stage_update'),
    ('DELETE', re.compile(r'^/api/admin/renewals/stages/\d+$'), 'renewal.stage_delete'),
    ('POST', re.compile(r'^/api/admin/renewals/rebuild$'), 'renewal.rebuild'),
    ('POST', re.compile(r'^/api/admin/renewals/\d+/move$'), 'renewal.move'),
    ('POST', re.compile(r'^/api/admin/renewals/\d+/comment$'), 'renewal.comment'),
    ('PATCH', re.compile(r'^/api/admin/renewals/\d+$'), 'renewal.update'),
    # teacher SPA
    ('POST', re.compile(r'^/api/submitLesson$'), 'lesson.submit'),
    # auth-мутации данных (2FA-поля Account меняет сам пользователь)
    ('POST', re.compile(r'^/api/auth/2fa/enable$'), 'account.twofa_enable'),
    ('POST', re.compile(r'^/api/auth/2fa/disable$'), 'account.twofa_disable'),
    ('POST', re.compile(r'^/api/auth/invite/accept$'), 'account.invite_accept'),
]

FALLBACK = 'other'


def resolve_operation(method: str, url: str) -> str:
    """Ключ операции по методу и пути. У вызывающего metadata['operation']
    имеет приоритет (например 'changelog.revert' проставляется сервисом)."""
    for rule_method, pattern, operation in RULES:
        if method == rule_method and pattern.match(url):
            return operation
    return FALLBACK


def rule_for_operation(operation: str):
    """Обратный поиск для фильтра ленты: operation → (method, regex) или None."""
    for rule_method, pattern, rule_operation in RULES:
        if rule_operation == operation:
            return rule_method, pattern
    return None
